File: src/content/docs/topics/cs/coding-problems/binary-search/074-search-a-2d-matrix-approach3.py
def search_matrix(matrix: list, target: int) -> bool:
    m, n = len(matrix), len(matrix[0])
    lo, hi = 0, m * n - 1
    while lo <= hi:
        mid = (lo + hi) // 2
        r, c = divmod(mid, n)
        if matrix[r][c] == target:
            return True
        if matrix[r][c] < target:
            lo = mid + 1
        else:
            hi = mid - 1
    return False

mat = [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]]
assert search_matrix(mat, 3) == True
assert search_matrix(mat, 13) == False
assert search_matrix([[1]], 1) == True
assert search_matrix([[1]], 2) == False
assert search_matrix([[1, 3]], 3) == True
assert search_matrix([[1], [3]], 1) == True
print("all tests pass")

File: src/content/docs/topics/cs/coding-problems/binary-search/074-search-a-2d-matrix-approach3.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function searchMatrix(matrix: number[][], target: number): boolean {
    const m = matrix.length, n = matrix[0].length;
    let lo = 0, hi = m * n - 1;
    while (lo <= hi) {
        const mid = (lo + hi) >> 1;
        const r = Math.floor(mid / n), c = mid % n;
        if (matrix[r][c] === target) return true;
        if (matrix[r][c] < target) lo = mid + 1;
        else hi = mid - 1;
    }
    return false;
}

const mat = [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]];
assert(searchMatrix(mat, 3) === true);
assert(searchMatrix(mat, 13) === false);
assert(searchMatrix([[1]], 1) === true);
assert(searchMatrix([[1]], 2) === false);
assert(searchMatrix([[1, 3]], 3) === true);
assert(searchMatrix([[1], [3]], 1) === true);
console.log("all tests pass");
